### PR TITLE
Aggregated timeseries nc4

### PR DESCRIPTION
**Pull request by diodon**
This is a complete new version of the code, using netCDF4 standard library for streaming aggregated data into the output file, with very low memory demand. As an example, aggregating TEMP from 650 WQM files will use less than 400MB of RAM. Still uses xarray to open and manipulate the input files, but pandas is no longer required.

### CODE REVIEW

**Comment by diodon** [comment]
in the last version the dependency on the package was removed (redundancy). Also added the timeseries_tools.py that contains a set of general functions used by the timeseries code generating scripts.

File: aodntools/timeseries_products/aggregated_timeseries.py [hunk 1]
```diff
@@ -319,231 +235,201 @@ def source_file_attributes(download_url_prefix, opendap_url_prefix):
...
+
+    ## create ncdf file, dimensions and variables
+    ds = Dataset(os.path.join(output_dir, outfile), 'w')
+    OBSERVATION = ds.createDimension('OBSERVATION', size=varlen_total)
+    INSTRUMENT = ds.createDimension('INSTRUMENT', size=n_files)
+
+    obs_float_template = {'datatype': 'float', 'zlib': True, 'dimensions': ('OBSERVATION'), "fill_value": 99999.0}
+    obs_byte_template = {'datatype': 'byte', 'zlib': True, 'dimensions': ('OBSERVATION'), 'fill_value': 99}
+    obs_int_template = {'datatype': 'int', 'zlib': True, 'dimensions': ('OBSERVATION')}
+    inst_S256_template = {'datatype': 'S256', 'dimensions': ('INSTRUMENT')}
```
Comment (mhidas) [inline aodntools/timeseries_products/aggregated_timeseries.py]: I thought we didn't need this anymore. Should the datatype be string?

File: aodntools/timeseries_products/aggregated_timeseries.py [hunk 2]
```diff
@@ -319,231 +235,201 @@ def source_file_attributes(download_url_prefix, opendap_url_prefix):
...
+    agg_variable = ds.createVariable(varname=var_to_agg, **obs_float_template)
+    agg_variable_qc = ds.createVariable(varname=var_to_agg + '_quality_control', **obs_byte_template)
+    DEPTH = ds.createVariable(varname='DEPTH', **obs_float_template)
+    DEPTHqc = ds.createVariable(varname='DEPTH_quality_control', **obs_byte_template)
+    PRES = ds.createVariable(varname='PRES', **obs_float_template)
+    PRESqc = ds.createVariable(varname='PRES_quality_control', **obs_byte_template)
+    PRES_REL = ds.createVariable(varname='PRES_REL', **obs_float_template)
+    PRES_RELqc = ds.createVariable(varname='PRES_REL_quality_control', **obs_byte_template)
+
+    TIME = ds.createVariable(varname='TIME', **obs_float_template)
```
Comment (mhidas) [inline aodntools/timeseries_products/aggregated_timeseries.py]: TIME, LATITUDE, LONGITUDE _should_ be of type `double`

File: aodntools/timeseries_products/aggregated_timeseries.py [hunk 3]
```diff
@@ -319,231 +235,201 @@ def source_file_attributes(download_url_prefix, opendap_url_prefix):
...
+    instrument_id = ds.createVariable(varname='instrument_id', **inst_S256_template)
+    LATITUDE = ds.createVariable(varname='LATITUDE', **inst_float_template)
+    LONGITUDE = ds.createVariable(varname='LONGITUDE', **inst_float_template)
+    NOMINAL_DEPTH = ds.createVariable(varname='NOMINAL_DEPTH', **inst_float_template)
+
+    ## main loop
+    for index, file in enumerate(files_to_agg):
+        with xr.open_dataset(os.path.join(input_dir, file)) as nc:
+            nc = in_water(nc)
+            file_variables = list(nc.variables)
```
Comment (mhidas) [inline aodntools/timeseries_products/aggregated_timeseries.py]: `file_variables` not used. Remove.

File: aodntools/timeseries_products/aggregated_timeseries.py [hunk 4]
```diff
@@ -1,113 +1,134 @@
...
+    """
+    Return a 1D array of 2D values
+    :param nc: dataset
+    :param varname: Variable of interest
+    :return: variable values flattened
+    """
+    return nc[varname].values.flatten()
+
+
+def get_instrumentID(nc):
```
Comment (mhidas) [inline aodntools/timeseries_products/aggregated_timeseries.py]: Should be `get_instrument_id` by convention

File: aodntools/timeseries_products/aggregated_timeseries.py [hunk 5]
```diff
@@ -319,231 +235,201 @@ def source_file_attributes(download_url_prefix, opendap_url_prefix):
...
-    encoding = {'TIME':                     {'_FillValue': False,
-                                             'units': time_units,
-                                             'calendar': time_calendar},
-                'LONGITUDE':                {'_FillValue': False},
-                'LATITUDE':                 {'_FillValue': False},
-                'instrument_id':            {'dtype': '|S256'},
-                'source_file':              {'dtype': '|S256'}}
-
-    write_netCDF_aggfile(agg_dataset, ncout_path, encoding)
+    data_code = get_data_code(var_to_agg)
```
Comment (mhidas) [inline aodntools/timeseries_products/aggregated_timeseries.py]: Let's add data code `Z` as well.

File: aodntools/timeseries_products/aggregated_timeseries.py [hunk 6]
```diff
@@ -319,231 +235,201 @@ def source_file_attributes(download_url_prefix, opendap_url_prefix):
...
+    varlen_list = [0] + varlen_list
+    varlen_total = sum(varlen_list)
+    n_files = len(files_to_agg)
+
+    ## create ncdf file, dimensions and variables
+    ds = Dataset(os.path.join(output_dir, outfile), 'w')
+    OBSERVATION = ds.createDimension('OBSERVATION', size=varlen_total)
+    INSTRUMENT = ds.createDimension('INSTRUMENT', size=n_files)
+
+    obs_float_template = {'datatype': 'float', 'zlib': True, 'dimensions': ('OBSERVATION'), "fill_value": 99999.0}
```
Comment (mhidas) [inline aodntools/timeseries_products/aggregated_timeseries.py]: In the output file, all the variables assigned to this template end up as type `double` - not sure why??

File: aodntools/timeseries_products/timeseries_tools.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
+## Utility functions for timeseries products
```
Comment (mhidas) [inline aodntools/timeseries_products/timeseries_tools.py]: Please remove this file from this PR. It isn't being used yet and is just adding more duplicated code. We can do the refactoring in a separate PR.

File: aodntools/timeseries_products/aggregated_timeseries.py [hunk 8]
```diff
@@ -1,113 +1,136 @@
...
-    if LATITUDE is a dimension has length 1
-    if LONGITUDE is a dimension has length 1
-
-    :param file: name of the netcdf file
+    Return list of errors found in the file:
+    Variables of interest are present
+    TIME. LATITUDE, LONGITUDE,  is present
+    NOMINAL_DEPTH is not present as variable or attribute
+    file_version is not FV01
+    if LATITUDE or LONIGUTDE dimension has length >1
```
[the remark this answers is not in the document]
Comment (diodon) [reply aodntools/timeseries_products/aggregated_timeseries.py]: OK, back to the previous version. lat/lon dims check added

File: aodntools/timeseries_products/aggregated_timeseries.py [hunk 9]
```diff
@@ -1,113 +1,136 @@
...
 
-
-def sort_files_to_aggregate(files_to_agg, input_dir=''):
+def sort_files(files_to_agg, input_dir=''):
     """
-    sort the list of files to aggregate by time_deployment start attribute
-
-    :param files_to_agg: list of file URLs
+    sort list of files according to deployment date
+    requires netcdf4, dateutil.parser as parse
```
[the remark this answers is not in the document]
Comment (diodon) [reply aodntools/timeseries_products/aggregated_timeseries.py]: fixed

File: aodntools/timeseries_products/aggregated_timeseries.py [hunk 10]
```diff
@@ -1,113 +1,136 @@
...
-    if 'NOMINAL_DEPTH' not in variables and 'instrument_nominal_depth' not in attributes:
-        error_list.append('no NOMINAL_DEPTH')
 
-    if VoI not in variables:
-        error_list.append(VoI + ' not in file')
+def get_variable_values(nc, variable):
+    """
+    Get value sof the variable and its QC flags.
+    If variable is not present, nan returned
+    If variable present but not its QC flags, QC set to 9
```
[the remark this answers is not in the document]
Comment (diodon) [reply aodntools/timeseries_products/aggregated_timeseries.py]: OK.

File: aodntools/timeseries_products/aggregated_timeseries.py [hunk 11]
```diff
@@ -1,113 +1,136 @@
...
+    :param nc: dataset
+    :param variable: name of the variable to get
+    :return: variable values and variable qc flags
```
[the remark this answers is not in the document]
Comment (diodon) [reply aodntools/timeseries_products/aggregated_timeseries.py]: fixed. returns always a numpy array

File: aodntools/timeseries_products/aggregated_timeseries.py [hunk 12]
```diff
@@ -1,113 +1,136 @@
...
-        # test the units. not used. To be implemented in the future
-        # if nc[VoI].attrs['units'] != variable_attribute_dictionary[VoI]['units']:
-        #     error_list.append('Wrong units: ' + nc[VoI].attrs['units'])
+        variable_values = np.repeat(np.nan, n_records)
+        variableQC_values = np.repeat(np.nan, n_records)
+
+    return variable_values, variableQC_values
+
+
+def get_varvalues(nc, varname):
```
[the remark this answers is not in the document]
Comment (diodon) [reply aodntools/timeseries_products/aggregated_timeseries.py]: removed

File: aodntools/timeseries_products/aggregated_timeseries.py [hunk 13]
```diff
@@ -319,231 +235,201 @@ def source_file_attributes(download_url_prefix, opendap_url_prefix):
...
-    ## main loop
+    varlen_list = []
+    bad_files = {}
+    rejected_files = []
 
-    ## get the variables attribute dictionary
-    with open(TEMPLATE_JSON) as json_file:
-        variable_attribute_dictionary = json.load(json_file)['_variables']
+    # default name for temporary file. It will be renamed at the end
+    outfile = str(uuid.uuid4().hex) + '.nc'
```
[the remark this answers is not in the document]
Comment (diodon) [reply aodntools/timeseries_products/aggregated_timeseries.py]: changed, but I need to rename the tmp file using shutil.move instead of os.rename to avoid the OSError: [Errno 18] Invalid cross-device link

File: aodntools/timeseries_products/aggregated_timeseries.py [hunk 14]
```diff
@@ -319,231 +235,201 @@ def source_file_attributes(download_url_prefix, opendap_url_prefix):
...
 
-    fileIndex = 0
-    rejected_files = []
-    bad_files = {}
-    applied_offset =[]      ## to store the PRES_REL attribute which could vary by deployment
+    ## sort the file list in chronological order
+    files_to_agg = sort_files(files_to_agg, input_dir=input_dir)
```
[the remark this answers is not in the document]
Comment (diodon) [reply aodntools/timeseries_products/aggregated_timeseries.py]: yes... changed.

File: aodntools/timeseries_products/aggregated_timeseries.py [hunk 15]
```diff
@@ -319,231 +235,201 @@ def source_file_attributes(download_url_prefix, opendap_url_prefix):
...
-    variableMainDF.index.rename('OBSERVATION', inplace=True)
-
-    ## get the list of variables
-    varlist = list(variableMainDF.columns) + list(variableAuxDF.columns)
-
-    ## set variable attributes
-    add_variable_attribute = {'PRES_REL': {'applied_offset_by_instrument': applied_offset}}
+        with xr.open_dataset(os.path.join(input_dir, file)) as nc:
+            ## clip to in water data only
+            nc = in_water(nc)
```
[the remark this answers is not in the document]
Comment (diodon) [reply aodntools/timeseries_products/aggregated_timeseries.py]: changed

File: aodntools/timeseries_products/aggregated_timeseries.py (hunk 1 above)
Comment (diodon) [reply aodntools/timeseries_products/aggregated_timeseries.py]: yes, datatype: str

File: aodntools/timeseries_products/aggregated_timeseries.py (hunk 2 above)
Comment (diodon) [reply aodntools/timeseries_products/aggregated_timeseries.py]: fixed. Also, the other dtypes set to float32

File: aodntools/timeseries_products/aggregated_timeseries.py (hunk 3 above)
Comment (diodon) [reply aodntools/timeseries_products/aggregated_timeseries.py]: removed

File: aodntools/timeseries_products/aggregated_timeseries.py (hunk 4 above)
Comment (diodon) [reply aodntools/timeseries_products/aggregated_timeseries.py]: fixed

File: aodntools/timeseries_products/aggregated_timeseries.py (hunk 5 above)
Comment (diodon) [reply aodntools/timeseries_products/aggregated_timeseries.py]: done

File: aodntools/timeseries_products/aggregated_timeseries.py (hunk 6 above)
Comment (diodon) [reply aodntools/timeseries_products/aggregated_timeseries.py]: fixed in a previous comment. Apparently netCDf library understands float as double. The correct dtype is np.float32

File: aodntools/timeseries_products/timeseries_tools.py (hunk 7 above)
Comment (diodon) [reply aodntools/timeseries_products/timeseries_tools.py]: removed

File: aodntools/timeseries_products/aggregated_timeseries.py [hunk 16]
```diff
@@ -319,231 +235,201 @@ def source_file_attributes(download_url_prefix, opendap_url_prefix):
...
+                bad_files.update({file: error_list})
+                rejected_files.append(file)
+
+
+    ## remove bad files form the list
+    for file in bad_files.keys():
+        files_to_agg.remove(file)
+
+
+    varlen_list = [0] + varlen_list
```
[the remark this answers is not in the document]
Comment (diodon) [reply aodntools/timeseries_products/aggregated_timeseries.py]: I need 0 as the start for the first slice of the netCDF variables. It could be done in a different way, but it will require some modifications to the logic of the code.
